### ISSUE TITLE
BufStream's AsyncBufRead impl requires the underlying stream to impl AsyncBufRead

### ISSUE THREAD

**Issue by thanatos** [tokio-rs/tokio#2106]
## Version
`tokio` version 0.2.9

## Platform

Linux

## Description

The [`AsyncBufRead` impl on `tokio::io::BufStream`](https://docs.rs/tokio/0.2.9/tokio/io/struct.BufStream.html#implementations) is the following:

```
impl<RW: AsyncBufRead + AsyncRead + AsyncWrite> AsyncBufRead for BufStream<RW>
```

That is, `BufStream`, in order to be `AsyncBufRead` requires the underlying stream (`RW`) to be `AsyncBufRead`. But that would seem to defeat the point of `BufStream`, which is to buffer an underlying `RW`…

This seems to be a trivial error in the trait bounds.

See also [this comment](https://github.com/tokio-rs/tokio/issues/2064#issuecomment-571715255). (But since I wasn't sure if that was 100% the problem in that issue, I decided it would be better to just open a more direct & clearer issue.)

**Comment by udoprog** [comment]
That definitely looks like a mistake. I've proposed relaxing it in #2108. Thanks.

**Comment by udoprog** [comment]
Fixed in #2108

**Comment by thanatos** [comment]
Thanks for taking the time to write that patch! I've tested tokio 0.2.11, and it appears that this is fixed.
